### ISSUE TITLE
Exporter spec editor: why do we limit the preview to 20 rows?

### ISSUE THREAD

**Issue by manuelma** [Spine-project/Spine-Toolbox#1267]
Do we tell the user that the output is going to be limited? If we do, I completely missed it and was puzzled for a while, checking back if I had the correct db. Ideally we would provide a spin-box or something so the user can select the limit and be aware of it.

But why the limit? Why not showing all the rows every time? I have been playing around with the new classes in spinedb_api trying to speed things up, and I think now it's fast enough. A mapping of thousands of rows is previewed in a few seconds (I commented out the code that limits the number of rows to see this). So if performance is the issue, I'd prefer to keep working on that and showing the full preview if possible.

**Comment by soininen** [comment]
The preview limits were indeed not shown anywhere. The new controls look very nice now and the limit for the number of tables might be handy too!

**Comment by PekkaSavolainen** [comment]
How about #923? Is that still a thing? Seems to be assigned to me 😬.

**Comment by manuelma** [comment]
That's right @PekkaSavolainen we can use the same approach there. Also, at the moment what we have in Exporter spec editor is a bit crazy in the sense that starts refetching rows as soon as you press the spinbox button or type a character in it. So we probably should use a QTimer (although the preview updater is quite smart in stopping 'obsolete' worker threads, we should avoid spawning threads as much as possible)

**Comment by soininen** [comment]
> Also, at the moment what we have in Exporter spec editor is a bit crazy in the sense that starts refetching rows as soon as you press the spinbox button or type a character in it. So we probably should use a QTimer (although the preview updater is quite smart in stopping 'obsolete' worker threads, we should avoid spawning threads as much as possible)

There is already a QTimer in Exporter's `PreviewUpdater` that starts the threads after 100ms.

**Comment by manuelma** [comment]
> There is already a QTimer in Exporter's PreviewUpdater that starts the threads after 100ms grace period.

I saw that, perhaps that's enough. So if I press the spinbox 5 times, it will start 5 threads and each of those (but the last) will be interrupted after 100 milliseconds?

**Comment by soininen** [comment]
> So if I press the spinbox 5 times, it will start 5 threads and each of those (but the last) will be interrupted after 100 milliseconds?

The threads will start only after 100ms has passed so if you press quickly enough you'll start only the last set of threads.

**Comment by manuelma** [comment]
I tested it and the experience is not so bad to be honest.

**Comment by soininen** [comment]
We can play around with the 100ms grace period if necessary. Maybe there is a sweeter spot?

**Comment by manuelma** [comment]
I think you can consistently make it crash if you press the spinbox really quickly. I wonder if we're trying to connect to a slot that's being executed or something like that...?

**Comment by soininen** [comment]
> I think you can consistently make it crash if you press the spinbox really quickly.

You have to click really quickly but indeed it crashes. I've opened a separate issue for that #1277

**Comment by soininen** [comment]
I believe with the preview limit controls implemeted we are done here.
